src/ray/core_worker/core_worker.cc

```diff
@@ -400,26 +402,29 @@ Status CoreWorker::Put(const RayObject &object, const ObjectID &object_id) {
 }
 
 Status CoreWorker::Create(const std::shared_ptr<Buffer> &metadata, const size_t data_size,
+                          const std::vector<ObjectID> contained_object_ids,
                           ObjectID *object_id, std::shared_ptr<Buffer> *data) {
   *object_id = ObjectID::ForPut(worker_context_.GetCurrentTaskID(),
                                 worker_context_.GetNextPutIndex(),
                                 static_cast<uint8_t>(TaskTransportType::RAYLET));
-  return Create(metadata, data_size, *object_id, data);
+  RAY_RETURN_NOT_OK(Create(metadata, data_size, *object_id, data));
+  // XXX: add contained
+  if (data) {
+    reference_counter_->AddOwnedObject(*object_id, GetCallerId(), rpc_address_);
+  }
+  return Status::OK();
 }
 
 Status CoreWorker::Create(const std::shared_ptr<Buffer> &metadata, const size_t data_size,
                           const ObjectID &object_id, std::shared_ptr<Buffer> *data) {
   return plasma_store_provider_->Create(metadata, data_size, object_id, data);
 }
 
-Status CoreWorker::Seal(const ObjectID &object_id, bool owns_object, bool pin_object) {
+Status CoreWorker::Seal(const ObjectID &object_id, bool pin_object) {
   RAY_RETURN_NOT_OK(plasma_store_provider_->Seal(object_id));
-  if (owns_object) {
-    reference_counter_->AddOwnedObject(object_id, GetCallerId(), rpc_address_);
-    if (pin_object) {
-      // Tell the raylet to pin the object **after** it is created.
-      RAY_CHECK_OK(local_raylet_client_->PinObjectIDs(rpc_address_, {object_id}));
-    }
+  if (pin_object) {
+    // Tell the raylet to pin the object **after** it is created.
+    RAY_CHECK_OK(local_raylet_client_->PinObjectIDs(rpc_address_, {object_id}));
```

Comment:
Hmm was this wrong before? You should pin the object even if you do not own it, right?

However, I don't think it is correct to remove the AddOwnedObject call either. Probably, we should just separate into 2 separate if blocks.

Reply:
The `AddOwnedObject` call was moved into `Create` because for task returns we only have access to the contained IDs during the `Create` call with the current code structure (I probably could have left that to a follow-up PR though).

I do think it was wrong before, we should consider whoever first creates an object to be its owner (not whoever generates the ID). This is not very consequential thought, it only matters for the case of random/custom object IDs, which aren't supported.

Reply:
Hmm I'm confused, the owner should be the one who generates the ID, right? Not necessarily the one who first creates the object value.

Also, I think I'm getting a bit confused now about why even have the AddOwnedObject calls in Seal/Create. It seems like we already have it in Put and for submitting a task, so I'm not sure why we need it in Seal/Create also? That is, wouldn't it be okay to just update the object with the contained IDs during Create with a separate call?

Reply:
`Put` is never called from Python, that's only used for Java. `ray.put()` in python calls `Create` and then `Seal` (to avoid a copy of the serialized object).

The case I was referring to about "the one who first creates the value" was when you do something like:
```
def f(id):
    ray.worker.global_worker.core_worker.put_object(1, object_id=id[0])

object_id = ray.ObjectID.from_random()
f.remote([object_id])
```

Clearly we shouldn't worry much about this though, as it isn't supported.

The state of the code now is that we add the owned object ID when an object is created during `ray.put()` but not when it's put into the object store as a task return. For the above case, it is also not added as an owned object ID or pinned.

Reply:
I see, thanks!